articles/121_realtime_proposal.md

```diff
@@ -0,0 +1,227 @@
+
+  - Inject alternative synchronization primitives from RTOS-specific APIs where applicable.
+
+As of the time of writing, the Executor and the IntraProcessManager in rclcpp are designed to partially follow the last guideline.
+However, the abstractions are messy (particularly for the IntraProcessManager).
+There is no real-time multithreaded Executor that uses "real-time safe" synchronization primitives, similarly there is no such IntraProcessManager--there is only the abstraction path that would allow such alternatives to be implemented.
+There are also synchronization primitives that are not hidden behind an abstraction layer for the global interrupt handler and for parameters.
+
+### Middleware selection and QoS
+
+If the underlying middleware used in ROS 2 is not suitable for real-time performance, the entire system will fail to meet the real-time requirement.
+Network communication and discovery, particularly asynchronous communication, introduces a nondeterministic element into software.
+Though some real-time operating systems partially mitigate this (e.g. disabling interrupts), careful selection and use of a middleware is important.
+
+The embedded component of the system proposed will likely use `freertps` as the underlying middleware.
+The "desktop" system can be matrix-tested using the various DDS implementations supported by ROS 2.
+This allows us to compare the real-time performance of the different DDS implementations.
+
+Quality of service is always an important consideration with DDS.
+Some DDS QoS options were designed with real-time performance in mind, such as the DEADLINE policy (see opendds.org, [QoS Policy Usages](http://www.opendds.org/qosusages.html)).
+It may be convenient to expose these policies and provide a "real-time" QoS profile in rmw as a suggestion to ROS 2 users.
+Benchmarking and profiling may provide data to support the introduction of more QoS options into the rmw API.
+
+### Component lifecycle
+
+It should be noted that the planned component lifecycle features are useful for real-time programming, since it is common that real-time systems have a initialization phase in which the real-time constraint is relaxed, in which it is safe to preallocate memory, change thread priority, etc.
+Distinguishing between initialization and execution could lead to better self-verification of real-time constraints during the execution phase (see the ROS 2 design article on [managed nodes](http://design.ros2.org/articles/node_lifecycle.html)).
+
+## Testing
+
+The proposed system must be tested for correctness and expected behavior.
+The performance requirements of the system, such as latency, must also be benchmarked, and the inner workings of the system must be accessible, in order to identify bottlenecks.
+Therefore we propose a mix of automated unit tests, performance benchmarking, and profiling/introspection tools.
+
+### Benchmarking
```

Comment:
Unfortunately I don't. I only heard about them through ROS Industrial. If their software is available for download you could probably produce a similar setup but it seems like it would take considerable effort. Still, I think it would be a valuable resource for maintaining the real-time suitability if ROS and nodes in the future. Something to do when more time is available?